### PR TITLE
WT-12610 Enable TSAN for src/log suppress specific functions only

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/log/log_slot.c [hunk 1]
```diff
@@ -40,10 +40,10 @@ __log_slot_dump(WT_SESSION_IMPL *session)
         __wt_errx(session, "    Release LSN: %" PRIu32 "/%" PRIu32, slot->slot_release_lsn.l.file,
           __wt_lsn_offset(&slot->slot_release_lsn));
         __wt_errx(session, "    Offset: start: %" PRIuMAX " last:%" PRIuMAX,
-          (uintmax_t)slot->slot_start_offset,
-          (uintmax_t)__wt_atomic_loadiv64(&slot->slot_last_offset));
+          (uintmax_t)__wt_atomic_loadi64(&slot->slot_start_offset),
+          (uintmax_t)__wt_atomic_loadi64(&slot->slot_last_offset));
```
Comment (luke-pearson) [inline src/log/log_slot.c:44]: I removed your v's from this variable as it's not volatile.

**Review by ajmorton — APPROVED** [review]
LGTM 👍 
Two small things to address but neither needs re-review.

File: src/log/log.c [hunk 2]
```diff
@@ -2483,7 +2483,7 @@ __wt_log_force_write(WT_SESSION_IMPL *session, bool retry, bool *did_work)
     WT_STAT_CONN_INCR(session, log_force_write);
     if (did_work != NULL)
         *did_work = true;
-    myslot.slot = log->active_slot;
+    myslot.slot = __wt_atomic_load_generic(&log->active_slot);
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/log/log.c:2486]: Yeah we need an atomic load pointer, and atomic store pointer.

File: test/evergreen/tsan_warnings.supp [hunk 3]
```diff
@@ -10,11 +10,10 @@ race:__wt_set_shared_double
 race:src/include/stat.h
 race:src/support/stat.c
 
-# FIXME-WT-12610 Fields like log->slot_state are composite variables and more complex than the usual storing 
-# and loading of ints. These fields will be investigated in a separate ticket. 
-race:src/log
+# FIXME-WT-12734 this function copies a 16 byte struct which needs further investigation.
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply test/evergreen/tsan_warnings.supp]: 05e483619

File: src/log/log_slot.c [hunk 4]
```diff
@@ -502,12 +504,16 @@ __wt_log_slot_destroy(WT_SESSION_IMPL *session)
      */
     for (i = 0; i < WT_SLOT_POOL; i++) {
         slot = &log->slot_pool[i];
+        /*
+         * The following two slot state loads occur during shutdown and therefore don't require
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/log/log_slot.c]: 05e483619

File: src/log/log.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/log/log.c]: 05e483619

**Comment by luke-pearson** [comment]
There's at least one more data race here to quash: 
```
==================
WARNING: ThreadSanitizer: data race (pid=206361)
  Read of size 4 at 0x7ba000004180 by thread T2:
    #0 __wt_spin_owned /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/include/mutex_inline.h:315:16 (libwiredtiger.so.11.3.0+0x15a3ee)
    #1 __wt_spin_unlock_if_owned /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/include/mutex_inline.h:325:9 (libwiredtiger.so.11.3.0+0x15a350)
    #2 __log_file_server /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_log.c:664:5 (libwiredtiger.so.11.3.0+0x158505)

  Previous write of size 4 at 0x7ba000004180 by main thread (mutexes: write M81):
    #0 __wt_spin_unlock /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/include/mutex_inline.h:207:19 (libwiredtiger.so.11.3.0+0x212461)
    #1 __wt_log_release /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/log/log.c:2023:9 (libwiredtiger.so.11.3.0+0x2172da)
    #2 __log_write_internal /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/log/log.c:2712:9 (libwiredtiger.so.11.3.0+0x21a96d)
    #3 __wt_log_write /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/log/log.c:2595:11 (libwiredtiger.so.11.3.0+0x21a006)
    #4 __wt_txn_checkpoint_log /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/txn/txn_log.c:538:9 (libwiredtiger.so.11.3.0+0x33e6b1)
    #5 __wt_connection_close /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_open.c:128:9 (libwiredtiger.so.11.3.0+0x15b0b5)
    #6 __conn_close /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_api.c:1291:5 (libwiredtiger.so.11.3.0+0x1358c5)
    #7 main /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/examples/c/ex_log.c:300:5 (ex_log+0x4b70ac)

  Location is heap block of size 15920 at 0x7ba000004000 allocated by main thread:
    #0 calloc /data/mci/cf02dab68c6203199a2c4dcdba60629c/toolchain-builder/tmp/build-llvm-v4.sh-Hnp/llvm-project-llvmorg/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:668:5 (ex_log+0x445887)
    #1 __wt_calloc /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/os_common/os_alloc.c:35:14 (libwiredtiger.so.11.3.0+0x25eebf)
    #2 __wt_logmgr_create /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_log.c:982:5 (libwiredtiger.so.11.3.0+0x157d85)
    #3 __wt_connection_workers /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_open.c:225:5 (libwiredtiger.so.11.3.0+0x15b7a8)
    #4 wiredtiger_open /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_api.c:3242:5 (libwiredtiger.so.11.3.0+0x134bd5)
    #5 main /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/examples/c/ex_log.c:258:5 (ex_log+0x4b6d48)

  Mutex M81 (0x7ba000004150) created at:
    #0 pthread_mutex_init /data/mci/cf02dab68c6203199a2c4dcdba60629c/toolchain-builder/tmp/build-llvm-v4.sh-Hnp/llvm-project-llvmorg/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:1258:3 (ex_log+0x44868d)
    #1 __wt_spin_init /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/include/mutex_inline.h:142:5 (libwiredtiger.so.11.3.0+0x158002)
    #2 __wt_logmgr_create /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_log.c:987:5 (libwiredtiger.so.11.3.0+0x157e0b)
    #3 __wt_connection_workers /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_open.c:225:5 (libwiredtiger.so.11.3.0+0x15b7a8)
    #4 wiredtiger_open /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_api.c:3242:5 (libwiredtiger.so.11.3.0+0x134bd5)
    #5 main /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/examples/c/ex_log.c:258:5 (ex_log+0x4b6d48)

  Thread T2 'log-close-serve' (tid=206391, running) created by main thread at:
    #0 pthread_create /data/mci/cf02dab68c6203199a2c4dcdba60629c/toolchain-builder/tmp/build-llvm-v4.sh-Hnp/llvm-project-llvmorg/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:965:3 (ex_log+0x446e5b)
    #1 __wt_thread_create /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/os_posix/os_thread.c:60:5 (libwiredtiger.so.11.3.0+0x26ca8f)
    #2 __wt_logmgr_open /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_log.c:1049:5 (libwiredtiger.so.11.3.0+0x15817e)
    #3 __wt_connection_workers /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_open.c:252:5 (libwiredtiger.so.11.3.0+0x15b81b)
    #4 wiredtiger_open /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/conn/conn_api.c:3242:5 (libwiredtiger.so.11.3.0+0x134bd5)
    #5 main /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/examples/c/ex_log.c:258:5 (ex_log+0x4b6d48)

SUMMARY: ThreadSanitizer: data race /data/mci/325575e22e25b4a19e349147e2918eac/wiredtiger/src/include/mutex_inline.h:315:16 in __wt_spin_owned

```

**Comment by luke-pearson** [comment]
This [4f8ce1e](https://github.com/wiredtiger/wiredtiger/pull/10429/commits/4f8ce1ef1b35e3b3bad8465c97a1060f08202786) change is kinda interesting, we certainly shouldn't be pushing atomic load and stores to every access of ID as its normally it's accessed within a lock. It's only these ownership checks that do it concurrently as far as I can tell.

**Comment by luke-pearson** [comment]
I also found this data race, which while not src/log is still coming up in my patch:
```
WARNING: ThreadSanitizer: data race (pid=61730)
  Write of size 8 at 0x7b9400000c98 by main thread (mutexes: write M188, write M13):
    #0 __wt_conn_dhandle_alloc /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/conn/conn_dhandle.c:238:5 (libwiredtiger.so.11.3.0+0x15088e)
    #1 __session_find_shared_dhandle /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/session/session_dhandle.c:775:5 (libwiredtiger.so.11.3.0+0x2e9add)
    #2 __session_get_dhandle /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/session/session_dhandle.c:805:5 (libwiredtiger.so.11.3.0+0x2e93a4)
    #3 __wt_session_get_dhandle /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/session/session_dhandle.c:886:9 (libwiredtiger.so.11.3.0+0x2e83dc)
    #4 __create_file /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/schema/schema_create.c:310:5 (libwiredtiger.so.11.3.0+0x2b0128)
    #5 __schema_create /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/schema/schema_create.c:1429:15 (libwiredtiger.so.11.3.0+0x2ae936)
    #6 __wt_schema_create /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/schema/schema_create.c:1489:11 (libwiredtiger.so.11.3.0+0x2ae518)
    #7 __create_colgroup /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/schema/schema_create.c:622:9 (libwiredtiger.so.11.3.0+0x2af796)
    #8 __schema_create /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/schema/schema_create.c:1427:15 (libwiredtiger.so.11.3.0+0x2ae91c)
    #9 __wt_schema_create /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/schema/schema_create.c:1489:11 (libwiredtiger.so.11.3.0+0x2ae518)
    #10 __wt_session_create /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/session/session_api.c:968:5 (libwiredtiger.so.11.3.0+0x2c49cf)
    #11 __session_create /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/session/session_api.c:1014:11 (libwiredtiger.so.11.3.0+0x2c9ad8)
    #12 main /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/examples/c/ex_col_store.c:442:5 (ex_col_store+0x4b6e82)

  Previous read of size 8 at 0x7b9400000c98 by thread T2 (mutexes: write M37):
    #0 __evict_clear_all_walks /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/evict/evict_lru.c (libwiredtiger.so.11.3.0+0x1f0470)
    #1 __evict_server /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/evict/evict_lru.c:439:15 (libwiredtiger.so.11.3.0+0x1efbe7)
    #2 __wt_evict_thread_run /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/evict/evict_lru.c:325:15 (libwiredtiger.so.11.3.0+0x1ef7f6)
    #3 __thread_run /data/mci/7865d05946da2f68ef1d2eca100ccc1b/wiredtiger/src/support/thread_group.c:31:9 (libwiredtiger.so.11.3.0+0x31c443)

```

**Comment by luke-pearson** [comment]
Given it's dhandle code I'd rather not touch it.

**Comment by ajmorton** [comment]
> This [4f8ce1e](https://github.com/wiredtiger/wiredtiger/pull/10429/commits/4f8ce1ef1b35e3b3bad8465c97a1060f08202786) change is kinda interesting, we certainly shouldn't be pushing atomic load and stores to every access of ID as its normally it's accessed within a lock. It's only these ownership checks that do it concurrently as far as I can tell.

Unless we can see a performance cost my preference is to make all accesses of a shared variable variable atomic. It brings us in line with an `_Atomic` implementation, or std::atomic<T> if we go the C++ route.

**Review by ajmorton — COMMENTED** [review]
Looks good, just waiting on the g++ `_Generic` fix.

**Review by ajmorton — APPROVED** [review]
LGTM!

**Pull request merged**